src/dtx/dtx_common.c

```diff
@@ -0,0 +1,528 @@
+			  &dth->dth_dte, 1,
+			  cont_hdl->sch_pool->spc_map_version);
+out:
+	D_DEBUG(DB_TRACE,
+		"Stop the DTX "DF_DTI" ver %u, dkey %llu, intent %s, "
+		"%s, %s, %s: rc = %d\n",
+		DP_DTI(&dth->dth_xid), dth->dth_ver,
+		(unsigned long long)dth->dth_dkey_hash,
+		dth->dth_intent == DAOS_INTENT_PUNCH ? "Punch" : "Update",
+		dth->dth_sync ? "sync" : "async",
+		dth->dth_non_rep ? "non-replicated" : "replicated",
+		dth->dth_leader ? "leader" : "non-leader", result);
+
+	if (dth->dth_leader && dth->dth_dti_cos != NULL)
+		D_FREE(dth->dth_dti_cos);
+	D_FREE_PTR(dth);
+
+	return result > 0 ? 0 : result;
+}
+
+int
+dtx_conflict(daos_handle_t coh, daos_unit_oid_t *oid, uuid_t po_uuid,
+	     uuid_t co_uuid, struct dtx_conflict_entry *dces, int count,
+	     uint32_t version)
+{
+	struct dtx_entry	*commit_dtes = NULL;
+	struct dtx_entry	*abort_dtes = NULL;
+	int			 commit_cnt = 0;
+	int			 abort_cnt = 0;
+	int			 rc = 0;
+	int			 i;
+
+	D_ALLOC_ARRAY(commit_dtes, count);
+	if (commit_dtes == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	D_ALLOC_ARRAY(abort_dtes, count);
+	if (abort_dtes == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	for (i = 0; i < count; i++) {
+		int	j;
+		bool	skip = false;
+
+		for (j = 0; j < i; j++) {
+			if (daos_dti_eq(&dces[i].dce_xid, &dces[j].dce_xid)) {
+				skip = true;
+				break;
+			}
+		}
+
+		if (!skip) {
+			rc = vos_dtx_lookup_cos(coh, oid, &dces[i].dce_xid,
+						dces[i].dce_dkey, true);
+			if (rc == -DER_NONEXIST)
+				rc = vos_dtx_lookup_cos(coh, oid,
+						&dces[i].dce_xid,
+						dces[i].dce_dkey, false);
+			if (rc == 0) {
+				daos_dti_copy(&commit_dtes[commit_cnt].dte_xid,
+					      &dces[i].dce_xid);
+				commit_dtes[commit_cnt++].dte_oid = *oid;
+				continue;
+			}
+
+			if (rc == -DER_NONEXIST) {
+				daos_dti_copy(&abort_dtes[abort_cnt].dte_xid,
+					      &dces[i].dce_xid);
+				abort_dtes[abort_cnt++].dte_oid = *oid;
+				continue;
+			}
+
+			goto out;
+		}
+	}
+
+	if (commit_cnt > 0)
+		rc = dtx_commit(po_uuid, co_uuid, commit_dtes, commit_cnt,
+				version);
+
+	if (abort_cnt > 0) {
+		rc = dtx_abort(po_uuid, co_uuid, abort_dtes, abort_cnt,
+				version);
+		if (rc == -DER_NONEXIST)
+			rc = 0;
+	}
+
+out:
+	if (commit_dtes != NULL)
+		D_FREE(commit_dtes);
+	if (abort_dtes != NULL)
+		D_FREE(abort_dtes);
+
+	return rc;
+}
+
+int
+dtx_batched_commit_register(struct ds_cont_hdl *hdl)
+{
+	struct ds_cont			*cont = hdl->sch_cont;
+	struct dtx_batched_commit_args	*dbca;
+	d_list_t			*head;
+
+	D_ASSERT(cont != NULL);
+
+	head = &dss_get_module_info()->dmi_dtx_batched_list;
+	d_list_for_each_entry(dbca, head, dbca_link) {
+		if (uuid_compare(dbca->dbca_cont_hdl->sch_cont->sc_uuid,
+				 cont->sc_uuid) == 0)
+			return 0;
+	}
+
+	D_ALLOC_PTR(dbca);
+	if (dbca == NULL)
+		return -DER_NOMEM;
+
+	cont->sc_closing = 0;
+	ds_cont_hdl_get(hdl);
+	dbca->dbca_cont_hdl = hdl;
+	d_list_add_tail(&dbca->dbca_link, head);
+
+	return 0;
+}
+
+void
+dtx_batched_commit_degister(struct ds_cont_hdl *hdl)
```

Comment:
degister?  What does that mean?   deregister?

It would be helpul to have some comments in this function.   In particular, we set flushed = 0 here and this ULT isn't responsible for setting flushed back to 1 but relies on it being set to return.   It would be helpful to describe what it's doing and what is expected flow a bit here.

Reply:
deregister, I will comment in the patch.